In [2]:
import pandas as pd
import numpy as np
import warnings
warnings.filterwarnings('ignore')
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity

In [3]:
df=pd.read_csv(r"C:\Binjus Ds\ml\books.csv")

In [4]:
df.head()

isbn13      isbn10           title subtitle  \
0  9780002005883  0002005883          Gilead      NaN   
1  9780002261982  0002261987    Spider's Web  A Novel   
2  9780006163831  0006163831    The One Tree      NaN   
3  9780006178736  0006178731  Rage of angels      NaN   
4  9780006280897  0006280897  The Four Loves      NaN   

                           authors                     categories  \
0               Marilynne Robinson                        Fiction   
1  Charles Osborne;Agatha Christie  Detective and mystery stories   
2             Stephen R. Donaldson               American fiction   
3                   Sidney Sheldon                        Fiction   
4              Clive Staples Lewis                 Christian life   

                                           thumbnail  \
0  http://books.google.com/books/content?id=KQZCP...   
1  http://books.google.com/books/content?id=gA5GP...   
2  http://books.google.com/books/content?id=OmQaw...   
3  http://books.google.com/books/content?id=FKo2T...   
4  http://books.google.com/books/content?id=XhQ5X...   

                                         description  published_year  \
0  A NOVEL THAT READERS and critics have been eag...          2004.0   
1  A new 'Christie for Christmas' -- a full-lengt...          2000.0   
2  Volume Two of Stephen Donaldson's acclaimed se...          1982.0   
3  A memorable, mesmerizing heroine Jennifer -- b...          1993.0   
4  Lewis' work on the nature of love divides love...          2002.0   

   average_rating  num_pages  ratings_count  
0            3.85      247.0          361.0  
1            3.83      241.0         5164.0  
2            3.97      479.0          172.0  
3            3.93      512.0        29532.0  
4            4.15      170.0        33684.0

In [5]:
df.shape

(6810, 12)

In [6]:
df.isnull().sum()

isbn13               0
isbn10               0
title                0
subtitle          4429
authors             72
categories          99
thumbnail          329
description        262
published_year       6
average_rating      43
num_pages           43
ratings_count       43
dtype: int64

In [8]:
#Treating 262 values in description
df["description"]=df["description"].fillna('')

In [9]:
df.isnull().sum()

isbn13               0
isbn10               0
title                0
subtitle          4429
authors             72
categories          99
thumbnail          329
description          0
published_year       6
average_rating      43
num_pages           43
ratings_count       43
dtype: int64

### TF-IDF Matrix

In [10]:
tfidf= TfidfVectorizer(stop_words='english')

In [11]:
tfidf_matrix=tfidf.fit_transform(df["description"])

In [12]:
tfidf_matrix

<6810x30338 sparse matrix of type '<class 'numpy.float64'>'
	with 216835 stored elements in Compressed Sparse Row format>

In [13]:
dense_array=tfidf_matrix.toarray()

In [15]:
#converting sparse matrix to array
dense_array

array([[0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.]])

In [16]:
tfidf_matrix.shape

(6810, 30338)

- The vectorizer  has created 30338 dimensional vector for the description for every book